ikarus/finiteelements/mechanics/enhancedassumedstrains.hh

```diff
@@ -171,10 +217,30 @@ protected:
         return Eigen::MatrixX<ScalarType>{};
     }();
 
+    auto strainFunction      = underlying().strainFunction(par, dx);
+    const auto C             = underlying().materialTangentFunction(par);
+    const auto geo           = underlying().localView().element().geometry();
+    const auto numberOfNodes = underlying().numberOfNodes();
+
     auto calculateMatrixContribution = [&]<typename EAST>(const EAST& easFunction) {
       typename EAST::DType D;
       calculateDAndLMatrix(easFunction, par, D, LMat, dx);
 
+      for (const auto& [gpIndex, gp] : strainFunction.viewOverIntegrationPoints()) {
+        const auto M            = easFunction.calcM(gp.position());
+        const double intElement = geo.integrationElement(gp.position()) * gp.weight();
+        const auto EVoigt       = strainFunction.evaluate(gpIndex, on(gridElement)).eval();
+        const auto CEval        = C(gpIndex);
+        auto stresses           = (CEval * M * alpha_).eval();
+        for (size_t i = 0; i < numberOfNodes; ++i) {
+          for (size_t j = 0; j < numberOfNodes; ++j) {
+            const auto kgIJ =
+                strainFunction.evaluateDerivative(gpIndex, wrt(coeff(i, j)), along(stresses), on(gridElement));
+            K.template block<Traits::mydim, Traits::mydim>(i * Traits::mydim, j * Traits::mydim) += kgIJ * intElement;
```

Comment:
The K coming from the displacment based element looks already like

`K = K_eu +K_G(S_u)` now you are adding another geometric stiffness `K = K_eu +K_G(S_u)+ K_G(S_alpha)` The one from the displacements should not be there, should it?

Reply:
I am not sure if I got the question right, but I can try to explain the implementation here. So basically, $\mathbf{K}_G$ is obtained by integrating ${\left( \frac{\partial^2 \mathbf{E}}{\partial \mathbf{d}^2} : \mathbf{S} \right)}$ over $\Omega$, with $\mathbf{S} = \mathbf{C} : \mathbf{E}$ and $\mathbf{E} = \mathbf{E}_u + \tilde{\mathbf{E}}$. Due to the additive decomposition, the part related to $\mathbf{E}_u$ is incorporated via `NonlinearElastic` and here only the contribution from $\tilde{\mathbf{E}}$ is included via [`auto stresses = (CEval * M * alpha_).eval();`](https://github.com/ikarus-project/ikarus/blob/572e568ba6707fd91055776c4623408f7d1e676e/ikarus/finiteelements/mechanics/enhancedassumedstrains.hh#L234) and further computing `K_G` out of it. `K_eu` is not modified and it comes from the displacement based formulation (`NonlinearElastic`). The subtracted term $\tilde{\mathbf{K}}$ due to the static condensation of the internal parameter $\alpha$ is still existing and its implementation is not modified. Is this clear? Is something wrong here then? Or do you mean `K_G(S_u)` should now not be there and should be set to zero?

Reply:
𝑆=𝐶:𝐸 is only correct for St.Venant Kirchoff. It should be computed from S=dW/dE evalulated at E=E_u+E_tilde, doesnt it?

Therefore the stress cannot be additively decomposed and the correct implementation would be K_g(S)

Reply:
@henrij22 and I also discussed this [here](https://github.com/ikarus-project/ikarus/pull/325#discussion_r1780692400) and we understood that due to the orthogonality condition (integration of $\mathbf{S}:\tilde{\mathbf{E}}$ should be zero), $\tilde{\mathbf{E}}$ doesn't contribute to $W$, but perhaps this is wrong? It only doesn't contributes to the total potential energy $\Pi$? If so, with this argument, $\mathbf{C} = \frac{\partial^2 W}{\partial \mathbf{E}^2}$ should also be modified as mentioned [here](https://github.com/ikarus-project/ikarus/pull/325#discussion_r1780692400)?

How will you then implement this? Subtract the `kg` part from the displacement based element and compute modified `kg` again in eas?

```cpp
auto stress_u           = underlying().getStress(EVoigt);
auto total_stress       = (stress_u + CEval * M * alpha_).eval();
for (size_t i = 0; i < numberOfNodes; ++i) {
    for (size_t j = 0; j < numberOfNodes; ++j) {
        const auto kgIJ_u =
                strainFunction.evaluateDerivative(gpIndex, wrt(coeff(i, j)), along(stress_u), on(gridElement));
        const auto kgIJ =
                strainFunction.evaluateDerivative(gpIndex, wrt(coeff(i, j)), along(total_stress), on(gridElement));
        K.template block<Traits::mydim, Traits::mydim>(i * Traits::mydim, j * Traits::mydim) +=
                (kgIJ - kgIJ_u) * intElement;
    }
```

Reply:
In my point of view several problems arise:
1. The current implementation of EAS variants only enhance the Green-Lagrangian strain. There, the decomposition looks like E=E_u+E_EAS(alpha). There, the contribution can be nicely seperated since the additions are additively added and the strains are _linear_ in terms of the new dofs  $$alpha$$. Therefore, the second derivative wrt. to  $$alpha$$ of the strains is zero and no geometric part arises. 
Therefore, $$\frac{\partial^2 E}{\partial \alpha^2} =0$$ and no geometric stiffness arises.

2. For EAS variants where the deformation gradient is enhanced, this is much more complicated and entangled since F=F_u+F_EAS(alpha),
would yield for the Green-Lagrangian strains $$E$$ a dependence up to _quadratic_ terms on $$alpha$$. Then, also 
a geometric contribution appears there. This might be different if the element is implemented in the two point setting with $$P:F$$.

3. If your displacement element is not implemented using $$S:E$$ you might have to do things differently and the enhancement might be not applicable.

4. If the deformation gradient is enhanced multiplicatively: $$F= F_u * F_{EAS}$$ then everything gets even more entangled and it might not be as simple as adding or substraction things subsequently. (This is not common so I think we don't have to consider that)

In the tests I only see that you check for St. Venant Kirchhoff, you might see problems for NeoHooke then due to the problems I stated above.

For an overview see: https://onlinelibrary.wiley.com/doi/epdf/10.1002/nme.1620330705

5. It might be better that the displacement-based element signals the setting in which it is set. Like Lagrangian or two-point and then inside the EAS method you can make choices.

Basically, you need from the displacement-based element:

A function to evaluate the stresses S and material tangent C at the enhanced stresses and the boperator and 
then you have to assemble the external forces and the stiffness matrix completely on your own without relaying on the already added contributions. You can use the stiffness matrix from displacement-based element, if you pass in the new stresses and new material tangent. Then you can modify it accordingly and perform static condensation.

The two references you gave are simplifying the situation a lot because they only use SVK and/or only use the total Lagrangian setting.

I would propose the following.
1. Refactor the displacement-based element that they have a new stiffness matrix function that can be called with a lambda to evaluate the stresses and material tangent for a given strain E. For these displacement-based elements these lambdas are trivial and the strain is the normal strain. From this, you can get the new "displacement-based part" of stiffness matrix and right-hand side.
2. The displacement-based element should signal which strain it provides when you call the strain method. Then you can dispatch at compile time, what you want or can enhance $$E,F,H$$. 
3. In the EAS element you can call the displacement-based element to get the b-operator and inside you can compute L,D,M and perform the static condenstation.
4. You don't have to impplement the other fancy enhancement methods but for now it is sufficient to only provided the $$S:E$$ based one.
5. Think about a mechanism how the skill Mixin does not first perform the displacement-based assembly, since we throw it out anyway(overwriting the contribution). This is only for performance for now. Setting it to zero might be enough to first implement the other points above.
6. Profit
7. 
Let me know if something is unclear

Reply:
For some reason, the tests are working at the moment with Neo-Hookean material model, but I assume that it is working as a special case for random element and random displacement configurations used in the AutoDiff test. It might break in certain cases. I will investigate it.

Regarding the theoretical aspects, I comment on my understandings below:
1. Enhancements to treat locking can be performed on any of the following: $E = E^C + \tilde{E}$ (Green-Lagrange strains - additive enhancement) or $H = H^C + \tilde{H}$ (displacement gradients - additive enhancement) or $F = \tilde{F}F^C$ (deformation gradient - multiplicative enhancement) or $F = F^C + \tilde{H}$ (deformation gradient - additive enhancement), etc.
2. When we enhance $E = E^C + \tilde{E}$, then this is a total Lagrangian formulation with PK2 stress ($S$) as the energetically conjugate stress measure. However, this formulation might lead to inconsistencies if would like to express our residuum using a two-point setting with other stress/strain measures, if we only use $F^C$ for push-forward/pull-back operations.
3. Alternatively, if we enhance $H$ or $F$, we can either express the fully-nonlinear version in a two point setting using perhaps PK1 stress ($P$) as the stress measure or have total-Lagrangian setting by computing $E$ using enhanced $H$ or $F$.
4. The current state of the PR doesn't capture $S$ and the tangent moduli $\mathbb{C}$ while computing $f_{int}$ and $K_{tot} = K_{eu} + K_g$ correctly as it doesn't evaluate them at $E = E^C + \tilde{E}$ correctly.

I mention my proposal for the current PR (ignoring the implementation aspects and only commenting on the objectives):
1. Fix the EAS element such that it address Point 4 (see above).
2. The implementation will only focus on total Lagrangian formulation mentioned in Point 2 (see above) based on $S$ and $E$.
(I will try to incorporate your points regarding the implementation aspects for the above two points)

Proposal for a separate PR:
1. After this PR is merged, I will work on Issue #85 and implement expressions for $H$ and $F$ in `dune-localfefunctions`.
2. Rework on the nonlinear finite elements such that it could also handle other nonlinear finite elements like Q1/Hm or Q1/HTm. Add a signal to indicate the type of formulation implemented (total Lagrangian or Two-Point) and the stress/strain measures used. I think the formulation type might also affect the skills `volume` and `traction` as it might also be necessary to integrate them in the same configuration as $f_{int}$ (I am not sure about this point, may be it is necessary only for update Lagrangian setting) and hence propose to do it separately.

@rath3t : What do you think about the proposal? Is something unclear or could be done differently? Please let me know.
@henrij22 : Do you have any other opinion here?